src/ILCompiler.ReadyToRun/src/ObjectWriter/SectionBuilder.cs

```diff
@@ -0,0 +1,964 @@
+
+            /// <summary>
+            /// Output stream to receive the relocated file
+            /// </summary>
+            private Stream _outputStream;
+
+            /// <summary>
+            /// Current position in the output file
+            /// </summary>
+            private int _outputFilePos;
+
+            /// <summary>
+            /// Relocation helper stores the output stream and initializes the PE blob builder enumerator.
+            /// </summary>
+            /// <param name="outputStream">Output stream for the relocated PE file</param>
+            /// <param name="peFileBuilder">PE file blob builder</param>
+            public RelocationHelper(Stream outputStream, ulong defaultImageBase, BlobBuilder peFileBuilder)
+            {
+                _outputStream = outputStream;
+                _outputFilePos = 0;
+                
+                _defaultImageBase = defaultImageBase;
+
+                _peFileLength = peFileBuilder.Count;
+                _peFileBlobs = peFileBuilder.GetBlobs();
+                FetchNextBlob();
+            }
+
+            /// <summary>
+            /// Copy data from the PE file builder to the output stream, stopping at given file position.
+            /// </summary>
+            /// <param name="filePos">Output PE file position to stop at</param>
+            public void CopyToFilePosition(int filePos)
+            {
+                CopyBytesToOutput(filePos - _outputFilePos);
+            }
+
+            /// <summary>
+            /// Copy all unprocessed data (after the last relocation) into the output file
+            /// without any further modifications.
+            /// </summary>
+            public void CopyRestOfFile()
+            {
+                do
+                {
+                    CopyBytesToOutput(_remainingLength);
+                }
+                while (TryFetchNextBlob());
+                
+                if (_outputFilePos != _peFileLength)
+                {
+                    // Input / output PE file length mismatch - internal error in the relocator
+                    throw new BadImageFormatException();
+                }
+            }
+
+            /// <summary>
+            /// Process a single relocation by copying the required number of bytes into a
+            /// buffer, applying the relocation and writing it to the output file.
+            /// </summary>
+            /// <param name="relocationType">Type of relocation to apply</param>
+            /// <param name="sourceRVA">RVA representing the address to relocate</param>
+            /// <param name="targetRVA">RVA representing the relocation target</param>
+            public void ProcessRelocation(ushort relocationType, int sourceRVA, int targetRVA)
+            {
+                Relocator relocator = Relocator.GetRelocator(relocationType);
+                int relocationLength = relocator.GetLength();
+                byte[] relocationBuffer = new byte[relocationLength];
```

Comment:
Same here: Allocating an array to process ever single little relocation is way too expensive.

Reply:
I have modified the logic to use a common preallocated buffer to avoid repeated allocations.